### ISSUE TITLE
unable to execute request: write: broken pipe

### ISSUE THREAD

**Issue by iredmail** [meilisearch/meilisearch-go#113]
Dear all,

I frequently got this error while adding document, any idea how to get detailed error? Is it caused by MeiliSearch server? Or the Go SDK?

**Comment by eskombro** [comment]
Hello @iredmail !

I haven't seen this problem before. Could you please tell us some steps to reproduce it? In which environment i syour MeiliSearch instance running?

Thanks for the info!

**Comment by iredmail** [comment]
Hi @eskombro,

Thanks for the reply.

- It's a 1 GB RAM (Linode) VPS for internal testing. It's a Ubuntu 20.04 box, runs MySQL (~= 30MB RAM), MeiliSearch (~= 290MB RAM), and 2 of our Golang programs (~= 15MB RAM each).
- Our Golang programs accepts and parse emails, then store parsed data in MeiliSearch.
- We generate ~= 20 emails per minute for testing, usually one of 10 may raises the error, randomly.
- MEILI_HTTP_PAYLOAD_SIZE_LIMIT is set to 100 MB, the testing email is just few KB.
- MEILI_ENV is "development".
- No any error in meilisearch log. Just the Golang program (which uses meilisearch-go package) raises the "broken pipe" error.

I have no code snippet to reproduce this issue, but our program can reproduce it each time. Any idea to help me locate the root cause? Thanks. :)

**Comment by eskombro** [comment]
Hey @iredmail 

Thanks for the info. I don't really know where this issue is coming from, but I am thinking that this could probably be related to the fasthttp client (we are using it since #88) dropping the connection. Do you have any Timeout error before the broken pipe?

If you have a moment, @senyast4745 can you see if you have any idea of what could be the problem?

**Comment by iredmail** [comment]
Hi @eskombro 

Will try version 0.12.1 and reply shortly. Thanks for helping.

**Comment by senyast4745** [comment]
Hello @iredmail , @eskombro.
If you don't mind, I can think about what happens in the evening

**Comment by iredmail** [comment]
Hi @senyast4745 

Yes i need some help. Thanks for helping. :)

**Comment by senyast4745** [comment]
Hi, @iredmail 

You can try to use your own fasthttp client, having previously eluded read \ write timeout from it.

**Comment by senyast4745** [comment]
@eskombro Perhaps we should start a task to add timeout configurations for our client

**Comment by iredmail** [comment]
I tried custom fasthttp client, but it didn't fix my issue. Maybe i didn't use correct settings?

```
fasthttpClient := &fasthttp.Client{
		ReadTimeout:         60 * time.Second,
		WriteTimeout:        60 * time.Second,
		MaxConnWaitTimeout:  60 * time.Second,
		MaxIdleConnDuration: 0 * time.Second,
		MaxConnDuration:     1 * time.Second, // close keepalive connection after 1 second
	}
```

**Comment by iredmail** [comment]
Tried 0.12.1, no more errors. WOW

**Comment by eskombro** [comment]
>     * Tried 0.12.1, no more errors. WOW
> 
>     * Binary file size of my program reduces 1 MB after version switch.

Ok, so I think this confirms that it is related to fasthttp but we will have to go a little deeper.

> Hello @iredmail , @eskombro.
> If you don't mind, I can think about what happens in the evening

So that would be amazing, thanks a lot :)


> I tried custom fasthttp client, but it didn't fix my issue. Maybe i didn't use correct settings?
> 
> ```
> fasthttpClient := &fasthttp.Client{
> 		ReadTimeout:         60 * time.Second,
> 		WriteTimeout:        60 * time.Second,
> 		MaxConnWaitTimeout:  60 * time.Second,
> 		MaxIdleConnDuration: 0 * time.Second,
> 		MaxConnDuration:     1 * time.Second, // close keepalive connection after 1 second
> 	}
> ```
> 
> Even worse, i now got a new error (the "broken pipe" error still exists):
> 
> > unable to execute request (path "PUT /indexes/messages/documents" with method "Documents.AddOrUpdate"): the server closed connection before returning the first response byte. Make sure the server returns 'Connection: close' response header before closing the connection

I would try using a bigger `MaxIdleConnDuration` since the connection is probably being dropped immediately as you set it to 0, cf: https://github.com/valyala/fasthttp/blob/master/client.go#L215-L219

**Comment by iredmail** [comment]
Dropping connection is on purpose (for testing).
I tried set `MaxIdleConnDuration` to 60 seconds before i tried 0.12.1, same error.

**Comment by iredmail** [comment]
Any update on the fix? :)

**Comment by senyast4745** [comment]
Hello everyone. I'm still out of town, so as soon as I get to the computer, I will definitely try to solve this problem.

**Comment by senyast4745** [comment]
@iredmail hello, can we go to the messenger, for example telegrams, so that I can respond more quickly and understand your problem in more detail?

**Comment by eskombro** [comment]
Hey @senyast4745 and @iredmail 

I would be interested in where this is coming from if you happen to find an explanation solution :)

Also, if you want MeilISearch Community Slack has a channel dedicated to Go SDK. If you feel comfortable with this, maybe it will be useful for another people to know what happens, so I invite you [here](https://meilicommunity.slack.com/)

**Comment by senyast4745** [comment]
@eskombro cool! May be you can create channel to discuss this issue?

**Comment by iredmail** [comment]
> @iredmail hello, can we go to the messenger, for example telegrams, so that I can respond more quickly and understand your problem in more detail?

Yes I’m available on telegram: the account name is @iredmail

**Comment by eskombro** [comment]
> @eskombro cool! May be you can create channel to discuss this issue?

It exists there already: https://meilicommunity.slack.com/archives/C01HCD33ZNK

**Comment by iredmail** [comment]
But I don’t use Slack. Also we already have this issue tracker for discussion, i don’t know why we need so many ways for communication. 

To reproduce the issue, just run a simple client and perform some AddOrUpdateDocument operations in short period, e.g. 20 http requests in one minute. That’s all. If it doesn’t occur, try to request more in one minute. 

The issue disappears if I switched to version 0.12.1.

**Comment by eskombro** [comment]
@iredmail 

Sure, this issue is more than enough to track the problem!

I wanted also to invite you both to this Slack channel if you ever want to discuss anything about the SDK :)

**Comment by senyast4745** [comment]
> But I don’t use Slack. Also we already have this issue tracker for discussion, i don’t know why we need so many ways for communication. 
> 
> To reproduce the issue, just run a simple client and perform some AddOrUpdateDocument operations in short period, e.g. 20 http requests in one minute. That’s all. If it doesn’t occur, try to request more in one minute. 
> 
> The issue disappears if I switched to version 0.12.1.

@iredmail ok i' ll try to reproduce your problem.

**Comment by iredmail** [comment]
Dear all,

Any update?

**Comment by senyast4745** [comment]
Hi, @iredmail 
I just ran tests. I was able to send 100 AddOrUpdate requests in 0.7 seconds, so I couldn't find your problem. Launched on Macbook 16 GB

**Comment by senyast4745** [comment]
My test

```package meilisearch

import (
	"fmt"
	"github.com/pkg/errors"
	"strconv"
	"sync"
	"testing"
)

func TestClientDocuments_AddOrUpdate_Load(t *testing.T) {
	var indexUID = "TestClientDocuments_AddOrUpdate"

	_, err := client.Indexes().Create(CreateIndexRequest{
		UID: indexUID,
	})

	if err != nil {
		t.Fatal(err)
	}

	var updatedIds []*AsyncUpdateID
	for i := 0; i < 50; i++ {
		updateIDRes, err := client.
			Documents(indexUID).
			AddOrUpdate([]docTest{
				{ID: strconv.Itoa(i) + "123", Name: "nestle" + strconv.Itoa(i)},
				{ID: strconv.Itoa(i) + "456", Name: "nestle" + strconv.Itoa(i)},
			})
		updatedIds = append(updatedIds, updateIDRes)
		if err != nil {
			t.Fatal(err)
		}

	}

	err = WaitingLongForUpdate(indexUID, updatedIds)
	if err != nil {
		t.Fatal(err)
	}

	var list []docTest
	err = client.Documents(indexUID).List(ListDocumentsRequest{
		Offset: 0,
		Limit:  100,
	}, &list)

	if err != nil {
		t.Fatal(err)
	}

	// tests are running in parallel so there can be more than 1 docs
	if len(list) < len(updatedIds) {
		t.Fatalf("number of doc should be at least %d, but actual %d", len(updatedIds), len(list))
	}
}

func WaitingLongForUpdate(indexUID string, updateIDs []*AsyncUpdateID) error {
	var wg sync.WaitGroup
	var err error
	for i := range updateIDs {
		wg.Add(1)
		i := i
		go func(wg *sync.WaitGroup) {
			defer wg.Done()
			var stat UpdateStatus
			stat, err = client.DefaultWaitForPendingUpdate(indexUID, updateIDs[i])
			if stat != UpdateStatusProcessed {
				err = errors.Errorf("Incorrect updated status %s. UpdateId %d", stat, i)
				fmt.Println(err)
			}
		}(&wg)
	}
	wg.Wait()
	if err != nil {
		return err
	}
	return nil
}

**Comment by eskombro** [comment]
Thanks for your check @senyast4745, I couldn't reproduce it either

**Comment by iredmail** [comment]
Please try to run Meilisearch on a 1GB RAM virtual machine, then test again.

**Comment by curquiza** [comment]
That's really weird, it looks like a MeiliSearch issue instead of the client issue.
MeiliSearch (and it's a known issue) cannot index a lot of documents in a small machine. 1GB might be not enough...
The core team is currently working on this issue by improving the current engine.
But it's still weird because you say you can use the go client v0.12.1 without any bug on this small machine...

**Comment by iredmail** [comment]
About the "engine", I have another "issue" with MeiliSearch: https://github.com/meilisearch/MeiliSearch/issues/1149

As @eskombro mentioned in [his reply](https://github.com/meilisearch/MeiliSearch/issues/1149#issuecomment-763498516), MeiliSearch requires "the same amount of RAM as the size the database takes on disk", this is too bad that i cannot use MeiliSearch as FTS for email archiving.

Is the core team working on improving this too?

**Comment by curquiza** [comment]
Yes the new core-engine will be way more performant.

**Comment by senyast4745** [comment]
> Но это все равно странно, потому что вы говорите, что можете использовать go client v0.12.1 без каких-либо ошибок на этой маленькой машине ...

It seems to me that the problem is that the v0.12.1 client also dropped out, just did it quietly.
@iredmail  Are you sure that all requests that you send work correctly on version v0.12.1

**Comment by iredmail** [comment]
> @iredmail Are you sure that all requests that you send work correctly on version v0.12.1

Yes.

**Comment by alallema** [comment]
I'm closing this issue because it is outdated. Feel free to reopen it if you need it.

**Comment by Pacerino** [comment]
Are there any updates?
I've been desperately trying to fix the error for hours. I am transferring all data from MySql to Meilisearch, but every few seconds I get the error message... It is frustrating.

**Comment by curquiza** [comment]
Hello @Pacerino!

Do you receive the exact same error message, I mean a broken pipe error? If yes, can you detail the exact steps you did, providing your dataset and your code if possible?

The problem here is we did not succeed to reproduce this error, so it's impossible to fix it. Your feedback could be really helpful.

If you do not get the exact same error message, please open another issue with your detailed steps.

**Comment by curquiza** [comment]
Thanks a lot for your detailed explanation @Pacerino!

It looks like your issue comes from MeiliSearch and not from this package specifically...

I can understand you don't have the time to check these two alternatives to your workaround, but I write it, it can help other people:
- MeiliSearch can accept big payloads (100Mb by default and [you can increase this value](https://docs.meilisearch.com/reference/features/configuration.html#payload-limit-size)). It means you can push way more than 5000 documents per batch, regarding the struct of `Items`.
- We are aware we have RAM consumption issues with MeiliSearch v0.20.0. This might leads to this error. v0.21.0 fixes part of this issue. It's going to be released soon, and meanwhile, you can check the [latest release candidate](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.21.0rc4). We are currently working on a new indexer that might be released before the end of September to fix the RAM consumption issue.

**Comment by Pacerino** [comment]
Hey @curquiza ,

Sorry for the late feedback!
The solutions that you have brought I already have tested. Unfortunately it brings nothing.

- My payload was at 1GB. The error still occurred.
- The server has 32GB of RAM and 6 cores. Meilisearch had the possibility to use the whole RAM and cores. Without success.

As I said before, the only option is to check if the data has already been indexed or not.

I am not sure if a similar error occurred with the other SDKs. However, it would confirm the assumption that the error lies with Meilisearch itself and not with the SDKs.

**Comment by curquiza** [comment]
Ok @Pacerino, good to know, thank you! :)

We did a huge refactor of MeiliSearch in v0.21.0. The core engine is not the same as the one in v0.20.0. It's really complicated for us to identify your issue right now, but maybe v0.21.0 has fixed this problem thanks to our refactor.

`meilisearch-go` is not yet compatible with v0.21.0 since the stable version of v0.21.0 is not out. I'll let you know here when you can test v0.21.0 with the newest version of this package, if you want to.

**Comment by iredmail** [comment]
@curquiza I know MeiliSearch uses LMDB, but any plan to support data larger than memory size? e.g. 100GB data on a machine with 8 GB memory.

**Comment by curquiza** [comment]
We improved the RAM consumption with v0.21.0 and we keep working on improving the indexer for v0.22.0. Currently, we still have crashes if we use too much memory, but this might be fixed in the future (v0.22.0 or v0.23.0, so August or September).
So, 8GB of RAM for a dataset of 100GB will take a lot of time to index, but it should work.

**Comment by curquiza** [comment]
Hello! [v0.21.0](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.21.0) is now out! :)
